### ISSUE TITLE
MSTest Runner: Provide a sample for usage in a project that uses WindowsAppSdk

### ISSUE THREAD

**Issue by abdes** [microsoft/testfx#2620]
It's not clear how the new MSTest runner fits in a context where the test project uses WindowsAppSdk or WinUI.

Such projects use a WinExe output type already and use custom setup code for the TestExecutor in the App startup code.

Please refer to: https://devblogs.microsoft.com/ifdef-windows/winui-desktop-unit-tests/ for more context on these types of test projects.

AB#2050328

**Comment by Evangelink** [comment]
Hi @abdes,

Thanks for the question! This was actually part of some of my experimentations yesterday :D

I have found and fixed a bug (see https://github.com/microsoft/testfx/pull/2617) and I was able to list what's working and not working. I need to sync with WinUI team to understand the way forward.

I will post a simple example later today, maybe you could review it and give me some feedback.

**Comment by abdes** [comment]
Thanks @Evangelink for the super fast reply. I'm always available if you need to try something.

**Comment by Evangelink** [comment]
Please find enclosed a basic example of supporting MSTest runner with WinUI project [MSTestRunnerWinUI.zip](https://github.com/microsoft/testfx/files/14695606/MSTestRunnerWinUI.zip). As stated in the csproj, not all modes are working (it seems that there is some issue with unpackage and self-contained).

**Comment by abdes** [comment]
@mit456 here is a Nuget.config file that will let you get the preview versions of MSTest:

```
<?xml version="1.0" encoding="utf-8"?>

<configuration>
    <!--
        Used to specify the default Sources for list, install and update.
        See: nuget.exe help list
        See: nuget.exe help install
        See: nuget.exe help update
    -->
    <packageSources>
        <clear />
        <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
        <add key="test-tools" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/test-tools/nuget/v3/index.json" />
    </packageSources>
    <!-- Define mappings by adding package patterns beneath the target source. -->
    <packageSourceMapping>
        <!-- key value for <packageSource> should match key values from <packageSources> element -->
        <packageSource key="test-tools">
            <package pattern="MSTest.*" />
            <package pattern="Microsoft.Testing.*" />
        </packageSource>
        <!-- key value for <packageSource> should match key values from <packageSources> element -->
        <packageSource key="nuget.org">
            <package pattern="*" />
        </packageSource>
    </packageSourceMapping>

    <!-- Used to disable package sources  -->
    <disabledPackageSources />

    <config>
        <!--
            Used to specify default source for the push command.
            See: nuget.exe help push
        -->
        <add key="defaultPushSource" value="https://api.nuget.org/v3/index.json" />
    </config>

    <packageRestore>
        <!-- Allow NuGet to download missing packages -->
        <add key="enabled" value="True" />

        <!-- Automatically check for missing packages during build in Visual Studio -->
        <add key="automatic" value="True" />
    </packageRestore>

    <packageManagement>
        <!-- Format is PackageReference -->
        <add key="format" value="1" />

        <!-- Do not show a prompt to chose the package format -->
        <add key="disabled" value="False" />
    </packageManagement>
</configuration>
```

Your .csproj seem OK. If you still run into problems, please look carefully at the sample provided by @Evangelink . Note that as of now, the setup works fine within Visual Studio Test Explorer, but I did not  find any way (yet), to make it work from the command line, either via the new executable runner or via `dotnet test` command, if the test cases use WinUI.

```
  Error Message:
   Exception thrown while executing test. If using extension of TestMethodAttribute then please contact vendor. Error message: UITestMethodAttribute.DispatcherQueue should not be null. To use UITestMethodAttribute within a WinUI Desktop App, remember to set the static UITestMethodAttribute.DispatcherQueue during the test initialization., Stack trace:    at Microsoft.VisualStudio.TestTools.UnitTesting.AppContainer.UITestMethodAttribute.Execute(ITestMethod testMethod) in /_/src/TestFramework/TestFramework.Extensions/Attributes/WinUI_UITestMethodAttribute.cs:line 70
   at Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.TestMethodRunner.ExecuteTest(TestMethodInfo testMethodInfo) in /_/src/Adapter/MSTest.TestAdapter/Execution/TestMethodRunner.cs:line 388
```

Even if we properly set the DispatcherQueue in the App.OnLaunched or in a Test Initializer.

**Comment by Evangelink** [comment]
For the time being, Test Explorer (both for VS and VS Code) will always run tests using VSTest platform and so won't use MSTest runner. We have provided an improvement to this but this is waiting to be shipped into VS.

> but when executing it from command line dotnet test, it is giving exact some error as above.

By default `dotnet test` will also fallback to VSTest platform. To enable the new runner mode with `dotnet test`, you will need to follow these instructions: https://learn.microsoft.com/dotnet/core/testing/unit-testing-platform-integration-dotnet-test#dotnet-test---microsofttestingplatform-mode

The support for WinUI with MSTest runner is not yet official as this is not something we have tested or worked on. Following @abdes suggestion, I have started to do some POC to see what's working or not. This is still considered as experimental but I do hope to find some time to sync with WinUI team and work on the failing parts so we can provide a proper experience.

**Comment by Evangelink** [comment]
There is indeed some bug with MSTest 3.2 for the support of WinUI. This is fixed in this PR #2617 that targets main (so v3.4) that I haven't backported to 3.3 that I will release this week because the support is not yet complete.

If you follow the steps described above for the nuget.config and use `3.4.0-preview.24171.3` or newer, you should not have this error.
